### PR TITLE
Nukes Couette and Lid-driven flows, and reintroduces figures for convergence tests

### LINKED ISSUE #1459: Validation + Convergence tests: images not showing up
See, e.g., the Stratified Couette flow validation section in Docs: 

Screen Shot 2021-03-13 at 4 44 11 pm

### LINKED ISSUE #1696: Validation Experiments in Docs
The Validation experiments in the Docs is very incomplete. I suggest we nuke it? Perhaps just leave the convergence tests page there and rename it to "Convergence Tests"?

### PR DESCRIPTION
**Pull request by tomchor**
The Convergence tests and  Couette flow pages (in validation) had missing figures that somehow got lost at some point. I was able to find the figures by going back to older versions of Oceananigans and reintroduced them.

Resolves #1459. Resolves #1696.

### CODE REVIEW

**Comment by navidcy** [comment]
How about #1696?

**Comment by tomchor** [comment]
> How about #1696? There was never any substantial discussion about it (other than a thumbs up ;)).

I wasn't aware of that one tbh. I'm okay with nuking everything except the convergence tests. But also we can approve and merge this one first (since it's obviously an improvement) and discuss what to do with the other validation after that. Both work for me.

**Comment by navidcy** [comment]
Let's see what @ali-ramadhan has to say.
If we are to nuke it then no point of bringing the pngs back on and adding to the git history :) We may as well nuke in the PR and rename the PR :)

**Comment by tomchor** [comment]
Okay, I took a second look at the other validation tests and I do think they should be kept. Sure, they're not super thorough, but I think they add to the docs. It's especially nice to see animations of the lid-cavity flow. So my vote goes to keeping them.

**Comment by glwagner** [comment]
The trade off is that it has to be maintained (as this PR shows). So the question is whether the maintenance cost is worth what they add.

**Comment by tomchor** [comment]
Does it need to be maintained in any way other than keeping that page up and making sure that it renders fine? If that's all it needs, then I'd say it's worth keeping.

Or are you suggesting that we keep re-running these tests with every minor Oceananigans release or something?

**Comment by glwagner** [comment]
Hmm. I think its value as a validation experiment really does depend on keeping it up to date with the current version of Oceananigans. What's really useful to people is the validation script that works for current Oceananigans that people can edit and modify, I think... Otherwise it's more a discussion that second-order finite volume / difference methods can reproduce the driven cavity, I suppose.

**Comment by tomchor** [comment]
> Hmm. I think its value as a validation experiment really does depend on keeping it up to date with the current version of Oceananigans. What's really useful to people is the validation script that works for current Oceananigans that people can edit and modify, I think... otherwise it's more a discussion that second-order finite volume / difference methods can reproduce the driven cavity, I suppose.
> 
> I understand the point though that this work has value and it's a shame to hide it away. I think it might be nice to have a separate repository of "cool setups", each of which links to the specific Oceananigans version that it works with...

I definitely see your point. But if that's true, then I think there's a deeper problem here which is that neither the validation scripts nor the benchmarks are kept up to date. (The benchmarks on the docs right now are done with v0.34.0!)

I actually tried discussing this in https://github.com/CliMA/Oceananigans.jl/issues/1634, and even proposed a solution that should work.

I think we have three options:

- We adopt the opinion that even though it's not ideal because it's out-of-date, we keep the validation experiments since they are cool set-ups that show that our code worked in the past (and supposedly should still be working now, although without proof)
- We adopt the opinion that if it's not up-to-date, it's not desired. So we nuke the benchmarks and the validation experiments altogether
- We try to fix this issue and keep the validation experiments and benchmarks up-to-date. You [seem to agree](https://github.com/CliMA/Oceananigans.jl/issues/1634#issuecomment-832203095) that creating a separate repo for validation and relying on `CompatHelper` to automatically create PRs would work. I'd be happy to help implementing that, but I don't have access to a server with GPUs that I set-up the CI for (or enough knowledge of the CI tbh).

Any one of those 3 options would work for me, but I think it would be good to make an explicit decision and stick to it, no?

**Comment by ali-ramadhan** [comment]
Just a note on this PR: It's kind of hacky but to avoid images and videos taking up space in git history (they take up a lot!) I'd advocate for uploading images into an issue (effectively hosting them on GitHub's servers) and then linking/embedding them.

In an ideal world the validation experiment CI pipeline would run and produce figures for all validation experiments which would then be automatically embedded into the docs with every version release. But setting up this infrastructure would take time and effort which I suspect no one can spare right now. Unfortunately it's a thankless job haha.

> We try to fix this issue and keep the validation experiments and benchmarks up-to-date.

I've always wanted to do this and wish I had the time to do it... This was the original motivitation for setting up the validation experiment CI pipeline: https://buildkite.com/clima/oceananigans-validation-experiments

It only runs the convergence tests right now and it still runs every night but it's been neglected so stuff is failing.

I personally think we should be reviving the validation experiments one-by-one starting with the most important ones by running them as part of that CI pipeline and fixing it when it fails. But without the ability to dedicate the time I can see why reducing our maintenance burden would be desirable.

I've also always wanted to set up a benchmarking CI pipeline, but I don't think we have the proper computational resources to set it up right now.

**Comment by glwagner** [comment]
I guess my opinion is that from an organizational and psychological point of view it makes the most sense to have "real" validation experiments and a validation pipeline _when / if_ somebody can take responsibility for maintaining it. At that time, I think that person will be showered with praise, making the job less thankless! I think this is the best strategy in the long run for eventually having legitimate validation experiments. Having them sitting around in a half-baked state could produce the false impression in a future collaborator that such a task does not remain to be done.

**Comment by ali-ramadhan** [comment]
Happy to open an issue (or post to #1634) when I have the time to work on CI.

What should we do with this PR? I think it's a net positive change so I'm happy to approve as long as the images aren't in git history. If we decide to nuke this part of the docs then I suppose this PR is moot and should be closed.

> _when_ / _if_ somebody can take responsibility for maintaining it.

I think responsibility for maintaining the pipeline should fall on all maintainers/developers, otherwise it's not sustainable. Ideally if you open a PR that breaks a validation experiment you should fix it. If done concurrently it should only consist of small changes so it should only be a small burden (although burdens to add up).

I guess we don't run the validation CI on every PR since it's too expensive so maintenance is tough right now. Maybe we can run validation CI before every tagged release or something? Better infrastructure is needed I suppose.

**Comment by glwagner** [comment]
> > _when_ / _if_ somebody can take responsibility for maintaining it.
> 
> I think responsibility for maintaining the pipeline should fall on all maintainers/developers, otherwise it's not sustainable.

Okay, what if change my language to "when / if somebody can take the lead for developing the functionality" ?

**Comment by francispoulin** [comment]
In terms of the validation experimens, from what I understand, they are created when someone does something new and the script is a proof by example, or sorts.  Since they are never run in tests they do become stale, and a bunch no longer run, as has been pointed out.  The same can be said for benchmarks.  

Examples are great because they are continually tested and therefore updated as needed.  There are a bunch, but still, I don't think as much is covered in the examples as we see in validation.  It would be nice for people to see whats in validation, but if they don't work then that causes other problems.

I don't claim there is a simple solution but I'm glad we're talking about it.

**Comment by ali-ramadhan** [comment]
Yeah I think some of those one-off experiments in `validation` should probably be nuked. Maybe we should put them in a `sandbox` directory in the future.

Ideally they would serve both as actual validation against existing results (e.g. convergence test, lid-driven cavity, viscous flow around a cylinder) and as advanced examples, so maybe they should become Literate examples (or Pluto examples/notebooks...).

**Comment by navidcy** [comment]
OK, let me put my 2 cents.

If we have the validations in the docs just because of the animations then we should just put the animations on YouTube or somewhere and point people there. Having the non-working versions of, e.g., `lid_cavity.jl` script in the repo and in the docs has lead many new users in the rabbit hole of figure it out. Then users feel that scripts are not working, sometimes they post an issue or ask in Slack. But who know how many other users are simply discouraged. Personally, if I download a code that includes non-working pieces of code I am very discouraged. Thus, bottom-line: I see @glwagner's point that the functionality or pipeline for maintaining healthy validations scripts (even a subset of them) then *only those* should be included in the Docs.

`Sandbox`: I'm so against a `sandbox` directory! It's like keeping your trash bin in your house for ever in case there is something useful there. I used to to it all the time: I'd keep commented out code in my scripts... But I think it's a bad practice. We need to let go. The repository and the tagged versions should not include half-bake scripts and what not. I'm happy with a sandbox repository `Oceananigans.jl-sandbox` where we keep things. Or a `sandbox` branch even.

Regarding this PR and what's should be kept in the Docs: I vote we keep the convergence tests in the docs and nuke the validations. I agree with the git history size issue and probably the method of including the `.png` files that  @ali-ramadhan suggests would work -- I've never done it though.

(btw, @ali-ramadhan, I've noticed an increase in the repo's size recently... :()

**Comment by ali-ramadhan** [comment]
Yeah I agree with no code being better than broken code. So I guess that means nuke non-working scripts and they can always be revived later with the intention of maintaining them long-term.

Also agree that sandbox is bad. We can discourage further short-term additions to `validation` and encourage people to use branches or other short-term repositories for them?

As for repo size yeah the biggest files are images and regression files which we could aim to get rid of in the long-term, maybe in one fell swoop with BFG repo cleaner. We've done in the past but it takes some preparation and is rewriting git history which is bad.

**Comment by tomchor** [comment]
> But who know how many other users are simply discouraged. Personally, if I download a code that includes non-working pieces of code I am very discouraged.

That's something very important to consider, I think, and I totally agree. It seems like we're converging to the "no code is better than broken code" perspective. Here's my pitch for this PR:

- Remove the Lid-drive cavity and Couette flow from the docs
- Keep the Convergence tests (maybe move it to the Appendix to make for a cleaner left menu)
- Remove the `validation` directory from the master branch as well (that's broken code for now and it prompts users to test them and get errors)
- Open another issue/PR to discuss what we'll do with the validation scripts (maybe another repo is a good idea)

Thoughts?

**Comment by glwagner** [comment]
I agree with docs changes. But it would disrupt a lot of important ongoing work to remove the `validation` branch from the repo without providing an alternative. So I think we should leave the `validation` scripts and discuss what to do next before removing them from master / main.

**Comment by tomchor** [comment]
> I agree with docs changes. But it would disrupt a lot of important ongoing work to remove the `validation` directory from the repo without providing an alternative place / way to collaborate on new features. So I think we should leave the `validation` scripts in the repo and discuss what to do next before removing them.

Is there ongoing work there? I thought all ongoing work was supposed to be confined to PRs. That's what I've been doing.

If that's true and we need to keep the `validation` dir, then at the very least we should make it obvious that those scripts aren't ready for users. The name "validation" kinda implies that users can use those scripts to validate their installation, so it's not a surprise that a lot of newcomers try those scripts out and get error. Maybe rename that dir to `work_in_progress`? And then put a `README.md` file in it explaining that most of the scripts are out-of-date?

**Comment by glwagner** [comment]
> Is there ongoing work there? I thought all ongoing work was supposed to be confined to PRs. That's what I've been doing.

Sometimes work spans multiple PRs, such as implementing a CubedSphere abstraction, or immersed boundaries. The validation experiments are being used for this work.

**Comment by navidcy** [comment]
@tomchor, let's do these in this PR:

> * Remove the Lid-drive cavity and Couette flow from the docs
> * Keep the Convergence tests (maybe move it to the Appendix to make for a cleaner left menu)

and open issue/other PR for `validation` directory.

File: docs/make.jl
Comment (navidcy) [inline docs/make.jl:122]: perhaps move the `convergence_tests.md` in the `appendix` directory?

**Review by navidcy — APPROVED** [review]
Before merging make sure that you **remove the `.png` files**.

File: docs/make.jl
Comment (navidcy) [reply docs/make.jl:122]: Also, comma needed at the end of previous line :)

**Comment by navidcy** [comment]
I'll let @ali-ramadhan explain exactly what's the best way to have the `.png`s there without burdening the git history.

**Comment by tomchor** [comment]
> I'll let @ali-ramadhan explain exactly what's the best way to have the `.png` for the convergence tests in without burdening the git history.

Meaning you're asking me to wait before merging?

**Comment by navidcy** [comment]
Yeap. @ali-ramadhan was suggesting we put the images on a GitHub issue instead `git add …`

**Comment by tomchor** [comment]
@ali-ramadhan do you have any suggestion for me to store the figures? Or can I merge this PR with the figures included?

Btw, the amount of figures I'm introducing is 4.4 MB worth. Which I wouldn't say is a lot. I can also always try to compress the images further and decrease their size.

**Comment by navidcy** [comment]
@tomchor, it adds up. We should try to reduce the repo file size. It’s already big.

**Comment by tomchor** [comment]
I removed the figures and used the links to them from an older Oceananigans version. Should be better (I think) than uploading the pictures to a PR and it doesn't add any pngs.

@navidcy is this good enough for merging? No pngs are being added anymore.

**Comment by navidcy** [comment]
That's indeed better. Sure, if tests pass and images appear in the docs preview then merge! :)

**Comment by tomchor** [comment]
Where do you get the link for the preview? I couldn't find it! (I know it probably hasn't been built yet, but I've tried after previous commits and couldn't see it anywhere)

**Comment by navidcy** [comment]
It's the same as the actual docs but you replace the `/dev/` or `/vX.X.X/` part with `/previews/PRXXXX/'.

https://clima.github.io/OceananigansDocumentation/previews/PR1797

**Pull request merged**
